cli/explain/compactgraph/testdata/generate.go

```diff
@@ -0,0 +1,224 @@
+	if buildWorkspaceDirectory == "" {
+		log.Fatalf("BUILD_WORKSPACE_DIRECTORY environment variable must be set, run with `bazel run`")
+	}
+	ownPackage := os.Args[1]
+	outDir := filepath.Join(buildWorkspaceDirectory, filepath.FromSlash(ownPackage))
+	logs, err := filepath.Glob(filepath.Join(outDir, "*.pb.zstd"))
+	if err != nil {
+		log.Fatalf("Failed to glob logs: %s", err)
+	}
+	for _, l := range logs {
+		if err := os.Remove(l); err != nil {
+			log.Fatalf("Failed to remove log: %s", err)
+		}
+	}
+
+	for _, tc := range []struct {
+		name         string
+		baseline     string
+		baselineArgs []string
+		changes      string
+		changedArgs  []string
+	}{
+		{
+			name:     "java_noop_impl_change",
+			baseline: JavaProject,
+			changes: `
+-- src/main/java/com/example/lib/Lib.java --
+package com.example.lib;
+
+public class Lib {
+    public static String getName() {
+      return "Lib"; // no-op impl change
+    }
+}
+`,
+		},
+		{
+			name:     "java_impl_change",
+			baseline: JavaProject,
+			changes: `
+-- src/main/java/com/example/lib/Lib.java --
+package com.example.lib;
+
+public class Lib {
+    public static String getName() {
+      return "Lib2"; // impl change
+    }
+}
+`,
+		},
+		{
+			name:     "java_header_change",
+			baseline: JavaProject,
+			changes: `
+-- src/main/java/com/example/lib/Lib.java --
+package com.example.lib;
+
+public class Lib {
+    public static String getName() {
+      return "Lib";
+    }
+
+    public static void foo() {}
+}
+`,
+		},
+		{
+			name: "env_change",
+			baseline: `
+-- MODULE.bazel --
+-- pkg/BUILD --
+genrule(
+    name = "gen",
+	outs = ["out"],
+	cmd = "env > $@",
+)
+`,
+			baselineArgs: []string{"--action_env=EXTRA=foo", "--action_env=OLD_ONLY=old_only", "--action_env=OLD_AND_NEW=old"},
+			changedArgs:  []string{"--action_env=NEW_ONLY=new_only", "--action_env=OLD_AND_NEW=new", "--action_env=EXTRA=foo"},
+		},
+		{
+			name: "non_hermetic",
+			baseline: `
+-- MODULE.bazel --
+-- pkg/BUILD --
+genrule(
+    name = "gen",
+	outs = ["out"],
+	cmd = "uuidgen > $@",
+)
+`,
+		},
+	} {
+		tmpDir, err := os.MkdirTemp("", "explain-test-*")
+		if err != nil {
+			log.Fatalf("Failed to create temp dir: %s", err)
+		}
+		defer os.RemoveAll(tmpDir)
+
+		extractTxtar(tmpDir, tc.baseline)
+		collectLog(tc.baselineArgs, tmpDir, filepath.Join(outDir, tc.name+"_old.pb.zstd"))
+
+		extractTxtar(tmpDir, tc.changes)
+		collectLog(tc.changedArgs, tmpDir, filepath.Join(outDir, tc.name+"_new.pb.zstd"))
+	}
+}
+
+func collectLog(args []string, projectDir, logPath string) {
+	outputBase, err := os.MkdirTemp("", "explain-testdata-*")
+	if err != nil {
+		log.Fatalf("Failed to create temp output base: %s", err)
+	}
+	defer os.RemoveAll(outputBase)
+	// Bazel's output base can contain files with no write permissions.
+	defer filepath.WalkDir(outputBase, func(path string, d fs.DirEntry, err error) error {
+		return os.Chmod(path, 0755)
+	})
+	cmd := exec.Command(
+		"bazel",
```

Comment:
Nit: assuming this is Bazelisk, it would be nice to get a `.bazelversion` setup here so we can reproduce/update these a bit easier in the future.